zst-volatile-macro/src/lib.rs

```diff
@@ -49,3 +55,35 @@ pub fn derive_volatile(tts: TokenStream) -> TokenStream {
     }
     .into()
 }
+
+fn parse_repr(attrs: &[Attribute]) -> Result<bool, &'static str> {
+    let repr = attrs
+        .iter()
+        .find(|a| a.path.is_ident("repr"))
+        .ok_or("VolatileStruct can only apply to #[repr(C, ...)]")?;
+
+    let repr_metas = match repr.parse_meta().unwrap() {
+        syn::Meta::List(l) => l,
+        _ => return Err("#[repr(...)] invalid format"),
+    };
+
+    // Make sure #[repr(C)] exits
+    repr_metas
+        .nested
+        .iter()
+        .find(|m| match m {
+            syn::NestedMeta::Meta(syn::Meta::Path(p)) => p.is_ident("C"),
+            _ => false,
+        })
+        .ok_or("VolatileStruct can only apply to #[repr(C, ...)]")?;
+
+    // Check if packed struct
+    Ok(repr_metas
+        .nested
+        .iter()
+        .find(|m| match m {
+            syn::NestedMeta::Meta(syn::Meta::Path(p)) => p.is_ident("packed"),
+            _ => false,
+        })
+        .is_some())
```

Comment:
At first I tried to add a new `P: Pack` generic parameter to `Volatile` where `Pack` is a trait bound with two dummy types implements it:

```Rust
pub trait Pack {}
pub enum Packed{}
pub enum UnPacked{}
impl Pack for Packed{}
impl Pack for Unpacked{}

pub struct Volatile<T, P: Packed, O = Zero> {
    _type_marker: PhantomData<T>,
    _offset_marker: PhantomData<O>,
    _pack_marker: PhantomData<P>,
}

impl<T, O> Volatile<T, Packed, O> {
    pub fn read(&self) -> T { .. }
    pub fn write(&mut self) -> T { .. }
}

impl<T, O> Volatile<T, UnPacked, O> {
    pub unsafe fn read_unaligned(&self) -> T { .. }
    pub unsafe fn write_unaligned(&mut self) -> T { .. }
}

```

But then I found it stills needs user to choose `Packed` or `Unpacked` when call `Volatile::from_ref(..)` ?

Reply:
Let's call this `Aligned` and `Unaligned`. This might possibly be useful even values which aren't part of a packed struct, but just happen to be unaligned.

`Volatile::from_ref` will always produce an aligned pointer because the reference it takes has to be aligned (unaligned references are not allowed in Rust). It's only the fields of a packed struct that can be unaligned.

I came up with a better way to read unaligned pointers than reading bytes and transmuting: We can create a wrapper type that's packed like so:
```rust
#[repr(C, packed)]
struct Unaligned<T> {
    inner: T,
}
```
We can use this type to read and write values even when they are not aligned because the wrapper type has no alignment requirements. This has the big advantage that we can avoid transmuting which equates to avoiding one copy operation.

We could possibly have a unique non-aligned wrapper type for each packed field. This would allow us to use the specific alignment of the struct and not always 1.